fenix/app/src/main/java/org/mozilla/fenix/home/topsites/TopSitesViewHolder.kt

```diff
@@ -32,9 +35,33 @@ class TopSitesViewHolder(
         val topSites =
             components.appStore.observeAsComposableState { state -> state.topSites }.value
                 ?: emptyList()
+        val wallpaperState = components.appStore
+            .observeAsComposableState { state -> state.wallpaperState }.value
+            ?: WallpaperState.default
+
+        var (titleTextColor, sponsoredTextColor, faviconCardBackgroundColor) = TopSiteColors.buildColors()
+
+        wallpaperState.composeRunIfWallpaperCardColorsAreAvailable { cardColorLight, cardColorDark ->
+            faviconCardBackgroundColor = if (isSystemInDarkTheme()) {
+                cardColorDark
+            } else {
+                cardColorLight
+            }
+        }
+
+        wallpaperState.currentWallpaper.textColor?.let {
```

Comment:
Seems like we are creating the `TopSiteColors` and then updating the local vars in this composable, we could maybe pass the `wallpaperAdaptedTextColor` to the buildColors function and get updated `TopSiteColors` object.

Reply:
I struggled with the suggested change quite a bit because I also need the `faviconCardBackgroundColor` from the above block. I couldn't fit everything quite as neatly as described, but I took an approach of refactoring the builder function to take the wallpaper state instead. This wasn't quite what you had in mind, but I am all for any new feedback.

Reply:
Made the suggested changes for `titleTextColor` and `sponsoredTextColor`. I left `faviconCardBackgroundColor` as is because of the the run block.